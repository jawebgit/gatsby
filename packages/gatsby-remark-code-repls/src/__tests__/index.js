jest.mock(`fs`, () => {
  return {
    existsSync: jest.fn(),
    readFileSync: jest.fn(),
  }
})

const fs = require(`fs`)
const Remark = require(`remark`)
const plugin = require(`../index`)

const {
  PROTOCOL_BABEL,
  PROTOCOL_CODEPEN,
} = require(`../constants`)

const REMARK_TESTS = {
  Babel: PROTOCOL_BABEL,
  Codepen: PROTOCOL_CODEPEN,
}

const remark = new Remark()

describe(`gatsby-remark-code-repls`, () => {
  beforeEach(() => {
    fs.existsSync.mockReset()
    fs.existsSync.mockReturnValue(true)

    fs.readFileSync.mockReset()
    fs.readFileSync.mockReturnValue(`const foo = "bar";`)
  })

  Object.keys(REMARK_TESTS).forEach(name => {
    describe(`${name} remark transform`, () => {
      const protocol = REMARK_TESTS[name]

      it(`generates a link for the specified example file`, async () => {
        const markdownAST = remark.parse(`[](${protocol}file.js)`)

        const transformed = plugin({ markdownAST }, { directory: `examples` })

        expect(transformed).toMatchSnapshot()
      })

      it(`generates a link with the specified target`, async () => {
        const markdownAST = remark.parse(`[](${protocol}file.js)`)

        const transformed = plugin({ markdownAST }, { directory: `examples`, target: `_blank` })

        expect(transformed).toMatchSnapshot()
      })

      it(`generates a link for files in nested directories`, async () => {
        const markdownAST = remark.parse(`[](${protocol}path/to/nested/file.js)`)

        const transformed = plugin({ markdownAST }, { directory: `examples` })

        expect(transformed).toMatchSnapshot()
      })

      it(`generates a link with the specified default text`, () => {
        const markdownAST = remark.parse(`[](${protocol}file.js)`)

        const transformed = plugin({ markdownAST }, { directory: `examples`, defaultText: `Click me` })

        expect(transformed).toMatchSnapshot()
      })

      it(`generates a link with the specified inline text even if default text is specified`, () => {
        const markdownAST = remark.parse(`[Custom link text](${protocol}file.js)`)

        const transformed = plugin({ markdownAST }, { defaultText: `Click me`, directory: `examples` })

        expect(transformed).toMatchSnapshot()
      })

      it(`verifies example files relative to the specified directory`, () => {
        const markdownAST = remark.parse(`[](${protocol}path/to/nested/file.js)`)

        plugin({ markdownAST }, { directory: `examples` })

        expect(fs.existsSync).toHaveBeenCalledWith(`examples/path/to/nested/file.js`)
      })

      it(`errors if you do not provide a directory parameter`, () => {
        const markdownAST = remark.parse(`[](${protocol}file.js)`)

        expect(() => plugin({ markdownAST })).toThrow(`Required REPL option "directory" not specified`)
      })

      it(`errors if you provide an invalid directory parameter`, () => {
        fs.existsSync.mockReturnValue(false)

        const markdownAST = remark.parse(`[](${protocol}file.js)`)

        expect(() => plugin({ markdownAST }, { directory: `fake` })).toThrow(`Invalid REPL directory specified "fake"`)
      })

      it(`errors if provided a link to a local file that does not exist`, async () => {
        fs.existsSync.mockImplementation(path => path === `examples`)

        const markdownAST = remark.parse(`[](${protocol}file.js)`)

        expect(() => plugin({ markdownAST }, { directory: `examples` })).toThrow(`Invalid REPL link specified`)
      })
    })
  })
})
