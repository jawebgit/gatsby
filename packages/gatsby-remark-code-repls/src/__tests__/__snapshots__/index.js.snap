// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-remark-code-repls Babel remark transform generates a link for files in nested directories 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"https://babeljs.io/repl/#?presets=react&code_lz=MYewdgzgLgBAZiEMC8MBEAjAhgJzQbiA\\" >",
        },
        Object {
          "type": "text",
          "value": "REPL",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 35,
          "line": 1,
          "offset": 34,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 35,
      "line": 1,
      "offset": 34,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Babel remark transform generates a link for the specified example file 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"https://babeljs.io/repl/#?presets=react&code_lz=MYewdgzgLgBAZiEMC8MBEAjAhgJzQbiA\\" >",
        },
        Object {
          "type": "text",
          "value": "REPL",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 20,
          "line": 1,
          "offset": 19,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Babel remark transform generates a link with the specified default text 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"https://babeljs.io/repl/#?presets=react&code_lz=MYewdgzgLgBAZiEMC8MBEAjAhgJzQbiA\\" >",
        },
        Object {
          "type": "text",
          "value": "Click me",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 20,
          "line": 1,
          "offset": 19,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Babel remark transform generates a link with the specified inline text even if default text is specified 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"https://babeljs.io/repl/#?presets=react&code_lz=MYewdgzgLgBAZiEMC8MBEAjAhgJzQbiA\\" >",
        },
        Object {
          "type": "text",
          "value": "Custom link text",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 36,
          "line": 1,
          "offset": 35,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 36,
      "line": 1,
      "offset": 35,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Babel remark transform generates a link with the specified target 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"https://babeljs.io/repl/#?presets=react&code_lz=MYewdgzgLgBAZiEMC8MBEAjAhgJzQbiA\\" target=\\"_blank\\" rel=\\"noreferrer\\">",
        },
        Object {
          "type": "text",
          "value": "REPL",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 20,
          "line": 1,
          "offset": 19,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Codepen remark transform generates a link for files in nested directories 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"/redirect-to-codepen/path/to/nested/file.js\\" >",
        },
        Object {
          "type": "text",
          "value": "REPL",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 37,
          "line": 1,
          "offset": 36,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 37,
      "line": 1,
      "offset": 36,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Codepen remark transform generates a link for the specified example file 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"/redirect-to-codepen/file.js\\" >",
        },
        Object {
          "type": "text",
          "value": "REPL",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 22,
          "line": 1,
          "offset": 21,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Codepen remark transform generates a link with the specified default text 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"/redirect-to-codepen/file.js\\" >",
        },
        Object {
          "type": "text",
          "value": "Click me",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 22,
          "line": 1,
          "offset": 21,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Codepen remark transform generates a link with the specified inline text even if default text is specified 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"/redirect-to-codepen/file.js\\" >",
        },
        Object {
          "type": "text",
          "value": "Custom link text",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 38,
          "line": 1,
          "offset": 37,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 38,
      "line": 1,
      "offset": 37,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-code-repls Codepen remark transform generates a link with the specified target 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "html",
          "value": "<a href=\\"/redirect-to-codepen/file.js\\" target=\\"_blank\\" rel=\\"noreferrer\\">",
        },
        Object {
          "type": "text",
          "value": "REPL",
        },
        Object {
          "type": "html",
          "value": "</a>",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 22,
          "line": 1,
          "offset": 21,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
